contracts/examples/adder/interact/src/basic_interact.rs

```diff
@@ -201,13 +209,43 @@ impl AdderInteract {
         println!("sum: {sum}");
     }
 
+    async fn fail_upgrade(&mut self, new_value: u32) {
```

Comment:
Instead of duplicating the code, just add an ExpectError parameter and a from parameter to the upgrade function and call it twice in the same test. Catch the error using .returns